e2e framework: replace custom cleanup support with Ginkgo v2 mechanisms

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

In contrast to ginkgo.AfterEach, ginkgo.DeferCleanup runs the callback in
first-in-last-out order. Using it makes the following test code work as
expected:

    f := framework.NewDefaultFramework("some test")

    ginkgo.AfterEach(func() {
        // do something with f.ClientSet
    })

Previously, f.ClientSet was already set to nil by the framework's cleanup code.

Fixes:  #111979

#### Does this PR introduce a user-facing change?
```release-note
e2e: tests can now register callbacks with ginkgo.BeforeEach/AfterEach/DeferCleanup directly after creating a framework instance and are guaranteed that their code is called after the framework is initialized and before it gets cleaned up. ginkgo.DeferCleanup replaces f.AddAfterEach and AddCleanupAction which got removed to simplify the framework.
```

pohly committed:
e2e framework: clean up instance after all other code ran

In contrast to ginkgo.AfterEach, ginkgo.DeferCleanup runs the callback in
first-in-last-out order. Using it makes the following test code work as
expected:

    f := framework.NewDefaultFramework("some test")

    ginkgo.AfterEach(func() {
        // do something with f.ClientSet
    })

Previously, f.ClientSet was already set to nil by the framework's cleanup code.

pohly commented:
I'm going to take this into https://github.com/kubernetes/kubernetes/pull/112043.

/close

pohly committed:
e2e framework: remove AddAfterEach

For cleanup purposes the ginkgo.DeferCleanup is a better replacement for
f.AddAfterEach:
- the cleanup only gets executed when the corresponding setup code ran
  and can use the same local variables
- the callback runs after the test and before the framework
  deletes namespaces (as before)
- if one callback fails, the others still get executed

For the original purpose (https://github.com/kubernetes/kubernetes/pull/86177 "This is
very useful for custom gathering scripts.") it is now possible to use
ginkgo.AfterEach because it will always get executed. Just beware that its
callbacks run in first-in-first-out order.

SataQiu reviewed (commented):
/lgtm

aojea commented:
/hold
I have some questions

aojea commented:
@pohly IIUIC we also don't need the `cleanupHandle` anymore 

It was removed some time ago https://github.com/kubernetes/kubernetes/pull/111063/ but since some storage tests override them I've asked to revert it to not have a partial state

aojea commented on test/e2e/framework/framework.go:
```diff
@@ -184,14 +183,18 @@ func NewFramework(baseName string, options Options, client clientset.Interface)
+	// remains valid as long as possible.
+	//
+	// In addition, AfterEach will not be called if a test never gets here.
+	ginkgo.DeferCleanup(f.AfterEach)
```
will this run twice? we add it to `	f.cleanupHandle = AddCleanupAction(f.AfterEach)` in the next line

aojea replied:
it sounds like the thing to remove is the `cleanupHandle` , seems that `DeferCleanup` just fixes https://github.com/onsi/ginkgo/v2/issues/222

pohly replied:
I think it removes itself once it ran once. But you are right, let's use this opportunity to also remove cleanupHandle.

pohly replied:
Done.

aojea commented on test/e2e/framework/framework.go:
```diff
@@ -103,11 +103,6 @@ type Framework struct {
 	cleanupHandle CleanupActionHandle
 
-	// afterEaches is a map of name to function to be called after each test.  These are not
-	// cleared.  The call order is randomized so that no dependencies can grow between
```
should we worry about moving from randomised to ordered?

pohly replied:
I don't think so. We are not breaking anything ("ordered" is a valid order when previously only "randomized" was guaranteed) and I am not worried about such dependencies actually occurring in practice. Even if they do, the Ginkgo model of deterministic execution would make them okay.

aojea commented:
can we have a small unit test that prevent future accidents changing the orders traversed?

pohly commented on test/e2e/framework/framework.go:
```diff
@@ -320,6 +301,20 @@ func (f *Framework) BeforeEach() {
 }
 
+func (f *Framework) dumpNamespaceInfo() {
+	if !ginkgo.CurrentSpecReport().Failed() ||
```
I don't have a strong preference. Usually I prefer less lines of code. :shrug: 

I can restore the original style.

aojea replied:
confirmed that it's style thing then ... but I find the long format has less chances of introducing bugs and more friendly for new users, avoiding the mental exercise of the `if` expression ... at the end the code is compiled and I expect to result on the same compiled code

pohly replied:
Reverted.

pohly commented:
> can we have a small unit test that prevent future accidents changing the orders traversed?

I was thinking about it, but testing framework.NewDefaultFramework in a unit test where framework.BeforeEach runs is tricky because it wants to connect to the apiserver. Let me check this again, perhaps we can fake it.

aojea commented:
> Let me check this again, perhaps we can fake it.

can't we use this one?

https://github.com/kubernetes/kubernetes/blob/67bde9a1023d1805e33d698b28aa6fad991dfb39/cmd/kube-apiserver/app/testing/testserver.go#L102

pohly commented:
[kubernetes/test/integration/framework/test_server.go](https://github.com/kubernetes/kubernetes/blob/67bde9a1023d1805e33d698b28aa6fad991dfb39/test/integration/framework/test_server.go#L62) looks promising and I tried using it, however a few things are missing:
- It wants to connect to etcd, which we don't have in a unit test. Is there a way to use an in-memory stub for etcd instead of a real instance?
- The framework needs a kube.config file which isn't available (or at least not readily available yet).

pohly commented:
We could of course do an integration tests for this, but that seems overkill.

pohly commented:
FYI, here's my current `./test/e2e/framework/internal/unittests/cleanup/cleanup_test.go` file. The output comparison will fail, but I am not even getting there :sweat_smile: 

<details>

```
/*
Copyright 2022 The Kubernetes Authors.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package cleanup

import (
	"testing"

	"github.com/onsi/ginkgo/v2"

	"k8s.io/kubernetes/test/e2e/framework"
	"k8s.io/kubernetes/test/e2e/framework/internal/output"
	integration "k8s.io/kubernetes/test/integration/framework"
)

// The line number of the following code is checked in TestFailureOutput below.
// Be careful when moving it around or changing the import statements above.
// Here are some intentionally blank lines that can be removed to compensate
// for future additional import statements.
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
//
// This must be line #50.

var _ = ginkgo.Describe("framework", func() {
	ginkgo.BeforeEach(func() {
		framework.Logf("before")
	})

	f := framework.NewDefaultFramework("foo")

	ginkgo.AfterEach(func() {
		framework.Logf("after")
		if f.ClientSet == nil {
			framework.Fail("Wrong order of cleanup operations: framework.AfterEach already ran and cleared f.ClientSet.")
		}
	})

	ginkgo.It("works", func() {
		ginkgo.DeferCleanup(func() {
			framework.Logf("cleanup")
		})
	})
})

func TestCleanup(t *testing.T) {
	_, _, tearDown := integration.StartTestServer(t, integration.TestServerSetup{})
	t.Cleanup(tearDown)

	expected := output.SuiteResults{
		output.TestResult{
			Name:   "framework works",
			Output: ``,
		},
	}

	output.TestGinkgoOutput(t, expected)
}
```

</details>

aojea commented:
give me some time, I know there is some helper somewhere that does it

aojea commented:
@pohly 

```diff
diff --git a/test/e2e/framework/internal/testing/integration_test.go b/test/e2e/framework/internal/testing/integration_test.go
new file mode 100644
index 00000000000..d5b0bd3a91d
--- /dev/null
+++ b/test/e2e/framework/internal/testing/integration_test.go
@@ -0,0 +1,43 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package test
+
+import (
+       "path"
+       "testing"
+
+       "github.com/google/uuid"
+       etcdserver "k8s.io/apiserver/pkg/storage/etcd3/testserver"
+       "k8s.io/apiserver/pkg/storage/storagebackend"
+       clientset "k8s.io/client-go/kubernetes"
+       kubeapiservertesting "k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
+       "k8s.io/kubernetes/test/e2e/framework"
+)
+
+func TestKubernetesE2EFramework(t *testing.T) {
+       cfg := etcdserver.NewTestConfig(t)
+       etcdClient := etcdserver.RunEtcd(t, cfg)
+       storageConfig := storagebackend.NewDefaultConfig(path.Join(uuid.New().String(), "registry"), nil)
+       storageConfig.Transport.ServerList = etcdClient.Endpoints()
+
+       server := kubeapiservertesting.StartTestServerOrDie(t, nil, []string{}, storageConfig)
+       defer server.TearDownFn()
+
+       f := framework.NewDefaultFramework("test")
+       f.ClientSet = clientset.NewForConfigOrDie(server.ClientConfig)
+
+}
```

pohly commented:
Do you also know how to get a normal kube.config file for the embedded apiserver? `f.ClientSet = clientset.NewForConfigOrDie(server.ClientConfig)` won't work because `framework.BeforeEach` will ignore it and fail.

If we have a file, then we can set `KUBECONFIG`. Modifying the environment is okay because the test that I posted above will run just a single test inside the test binary.

aojea commented:
does it ignore it? I see it checks if is nil

```go
// BeforeEach gets a client and makes a namespace.
func (f *Framework) BeforeEach() {
	f.beforeEachStarted = true

	// The fact that we need this feels like a bug in ginkgo.
	// https://github.com/onsi/ginkgo/v2/issues/222
	f.cleanupHandle = AddCleanupAction(f.AfterEach)
	if f.ClientSet == nil {
```


anyway, I think you can reuse this

https://github.com/kubernetes/kubernetes/blob/67bde9a1023d1805e33d698b28aa6fad991dfb39/test/integration/examples/apiserver_test.go#L412-L447

also, @soltysh pointed me to 

> here are a few helpers in [k8s.io/client-go/tools/clientcmd/loader.go](http://k8s.io/client-go/tools/clientcmd/loader.go)  or [k8s.io/client-go/tools/clientcmd/config.go](http://k8s.io/client-go/tools/clientcmd/config.go) but most are working with clientcmd.Config

pohly commented:
> does it ignore it?

You are right, it doesn't. But then it also skips a bunch of code that normally gets executed, so let's try to run it "the normal way".

I'll check out the pointers that you provided, thanks.

pohly committed:
e2e internal: support Ginkgo v2 log time stamps

When using By or some other Ginkgo output functions, Ginkgo v2 now adds a time
stamp at the end of the line that we need to ignore. Will become relevant when
testing more complete output.

pohly committed:
test: add helper code for running apiserver inside unit test

This runs etcd and an apiserver using it inside the test process. The caller
can either use the ClientSet or the config file. More options might get added
in the future.

Co-author: Antonio Ojea <antonio.ojea.garcia@gmail.com>

pohly committed:
e2e framework: add unit test for test execution

This covers multiple facets of the current framework and of Ginkgo:

- Ginkgo output is verbose and includes detailed progress
  messages (BeforeEach/AfterEach tracing).
- Namespace creation.
- Order of callback invocation.

pohly committed:
e2e: remove cleanup actions

The framework.AddCleanupAction API was a workaround for Ginkgo v1 not invoking
AfterEach callbacks after a test failure. Ginkgo v2 not only fixed that, but
also added a DeferCleanup API which can be used to run some code if (and only
if!) the corresponding setup code ran. In several cases that makes the test
cleanup simpler.

pohly commented:
I've pushed a version of the commits with:
- a new unit test
- AddCleanupAction removed

pohly commented on test/e2e/framework/internal/unittests/testapiserver.go:
```diff
@@ -0,0 +1,131 @@
+	KubeConfigFile string
+}
+
+func StartAPITestServer(t *testing.T) TestAPIServer {
```
I am undecided whether this helper should be here or somewhere else, like `test/utils/apiserver`. Therefore I also haven't written any documentation for it.

Any thoughts?

pohly replied:
I moved it to `test/utils/apiserver`. That also made `verify-import-boss.sh` happy, although I don't fully understand why depending on the same packages indirectly is fine when depending on them directly is not.

It doesn't matter for `test/e2e/framework` because the goal no longer is to move it to staging.

pohly commented on test/e2e/framework/internal/unittests/testapiserver.go:
```diff
@@ -0,0 +1,131 @@
+	config := clientcmdapi.NewConfig()
+
+	credentials := clientcmdapi.NewAuthInfo()
+	credentials.Token = clientCfg.BearerToken
```
Almost exactly the same `createKubeConfig` function also exists in https://github.com/kubernetes/kubernetes/blob/23790ec7fa8fd3c5b18edd85bc12ba4d76f50e1f/test/e2e/framework/test_context.go#L427-L463.

The only difference is that the other one also sets
```
 	credentials.TokenFile = clientCfg.BearerTokenFile
```

If we keep this helper under test/e2e, then I should move that function into a common location.

pohly replied:
I removed the function from `test_context.go` and made sure that the copy in the new `test/utils/apiserver` package is identical (i.e. with `credentials.TokenFile = clientCfg.BearerTokenFile`).

pohly committed:
e2e framework: use new apiserver helper package

pohly commented:
The unit test triggered a data race in klog -> https://github.com/kubernetes/klog/issues/348.

pohly commented:
/retest

pull-kubernetes-node-e2e-containerd: `failed to run SSH command: out: ssh_exchange_identification: Connection closed by remote host`

aojea commented on test/e2e/framework/framework.go:
```diff
@@ -179,7 +174,6 @@ func (f *Framework) BeforeEach() {
 
 	// The fact that we need this feels like a bug in ginkgo.
 	// https://github.com/onsi/ginkgo/v2/issues/222
```
nit, the comment no longer applies

pohly replied:
Right. This begs the question: should I keep the if check at all?

The only situation where it is not nil is in the situation that you brought up recently elsewhere:
```
f := framework.NewDefaultFramework
f.ClientSet = <some client set>
```

If there's exactly one `It` after that, then it will use this client set provided by the setup code. But if there's a second `It`, then the framework.AfterEach will clear it and framework.BeforeEach will set it up anew.

Also, in a normal Ginko suite, where is this `<some client set>` supposed to come from? The code above runs at init time, before command line parsing. That makes it impossible to configure how to connect to a cluster.

Overall I think we should not support such code and simply remove the if check.

pohly replied:
I have pushed a version that drops the obsolete comment and the if check.

aojea replied:
yeah, I don't quite understand the logic it seems it is 
1. start test - create clientset
2. run test
3. end test, nil clientset

In theory, I think we should create the clientset once we start and reuse it ... that is BeforeSuite IIUIC 
https://onsi.github.io/ginkgo/#suite-setup-and-cleanup-beforesuite-and-aftersuite

anyway, I think your reasoning is correct and this was this way before :shrug:

pohly replied:
The clientset might be using different options for each test, therefore it has be created anew:
https://github.com/kubernetes/kubernetes/blob/2b2be7fa6b69c7591a31475fc4fb1dcff362a128/test/e2e/framework/framework.go#L204-L205

I also prefer it that way because then tests are better separated from each other.

aojea commented on test/e2e/storage/flexvolume_mounted_volume_resize.go:
```diff
@@ -109,24 +104,12 @@ var _ = utils.SIGDescribe("[Feature:Flexvolumes] Mounted flexvolume expand[Slow]
-	ginkgo.AfterEach(func() {
-		framework.Logf("AfterEach: Cleaning up resources for mounted volume resize")
-
-		if c != nil {
```
naive question, do you still need to check if is nil?

pohly replied:
No. Because the code now runs only after c was set, it can never be nil.

aojea commented on test/e2e/framework/framework.go:
```diff
@@ -169,33 +155,25 @@ func NewFramework(baseName string, options Options, client clientset.Interface)
 		Timeouts:                 NewTimeoutContextWithDefaults(),
 	}
 
-	f.AddAfterEach("dumpNamespaceInfo", func(f *Framework, failed bool) {
-		if !failed {
-			return
-		}
-		if !TestContext.DumpLogsOnFailure {
-			return
-		}
-		if !f.SkipNamespaceCreation {
-			for _, ns := range f.namespacesToDelete {
-				DumpAllNamespaceInfo(f.ClientSet, ns.Name)
-			}
-		}
-	})
-
 	ginkgo.BeforeEach(f.BeforeEach)
-	ginkgo.AfterEach(f.AfterEach)
```
technical question, should 
```
	// In addition, AfterEach will not be called if a test never gets here.
	ginkgo.DeferCleanup(f.AfterEach)

	// Registered later and thus runs before deleting namespaces.
	ginkgo.DeferCleanup(f.dumpNamespaceInfo)
```

go here , in `NewDefaultFramework`  or inside `BeforeEach()`

The constructor seems the one configuring the hooks order, or do we want the dependency that f.AfterEach only happens if f.BeforeAch has called?

pohly replied:
It has to be in `BeforeEach` because `DeferCleanup` is not allowed outside of a `BeforeEach/AfterEach/It` node. That is a somewhat artificial limitation, but conceptually it makes sense: it is meant to be used in combination with some setup code if and only if that setup code actual ran.

That makes sense for us here. If `f.BeforeEach` never gets reached, for example because some earlier `BeforeEach` skipped the test, then there is no need to run `f.AfterEach`.

pohly commented:
Oh how ironic, conflicting with my own changes to `hack/.import-aliases`. I should do less work.

dims commented:
LOL @pohly

pohly commented:
/hold

The klog data race caused pull-kubernetes-unit  to fail again. Let's update klog first.

aojea commented on test/e2e/framework/internal/unittests/cleanup/cleanup_test.go:
```diff
@@ -0,0 +1,151 @@
+`
+)
+
+func TestCleanup(t *testing.T) {
```
this test is excellent

mauriciopoppe replied:
this is cool!

aojea commented on test/e2e/storage/flexvolume_online_resize.go:
```diff
@@ -71,15 +70,11 @@ var _ = utils.SIGDescribe("[Feature:Flexvolumes] Mounted flexvolume volume expan
+		nodeLabelValue = ns
+		nodeKeyValueLabel = map[string]string{nodeKey: nodeLabelValue}
+		framework.AddOrUpdateLabelOnNode(c, nodeName, nodeKey, nodeLabelValue)
+		ginkgo.DeferCleanup(framework.RemoveLabelOffNode, c, nodeName, nodeKey)
```
this DeferCleanup seems to be superuseful now that I see how you are using it

aojea commented:
lgtm
/approve

the test fails with the reported data race in klog

> The unit test triggered a data race in klog -> [kubernetes/klog#348](https://github.com/kubernetes/klog/issues/348).

aojea commented:
/lgtm

the hold will wait for the klog fix

pohly commented:
@aojea: I took out the [.import-aliases](https://github.com/kubernetes/kubernetes/pull/111998/files#diff-ba0e2e4ba18d6e4284cd0ed68656beb5ef0f8fe804b67393a682c1f90426ae35) change, so now we no longer need additional approvals. Can you re-add your LGTM?

aojea commented:
/lgtm

aojea commented:
/test pull-kubernetes-integration

> TestDSCUpdatesPodLabelAfterDedupCurHistories

:eye:

aojea commented:
/hold cancel

the dependency on klog fix merged https://github.com/kubernetes/kubernetes/pull/112349
